### ISSUE TITLE
Windows 11 ARM64 Processor x86, x64 COM API Hook Crash.

### ISSUE THREAD

**Issue by kimjw0820** [microsoft/Detours#292]
Windows 11 ARM64 Processor x86, x64 COM Hook Crash.
We tested using the latest sources. (main - 2022-08-16 commit)

On Windows 11 ARM64, x64 processes are emulated. Therefore, ARM, ARM64, x86, and x64 processes operate in the ARM64 environment. There is no problem with win32 API Hooking of x86 and x64 processes in the ARM64 environment.
![image](https://github.com/microsoft/Detours/assets/6908230/3bc837ac-bfff-49d9-9073-e8f4b4c426b1)

The problem occurs when hooking the COM API of x86 and x64 processes in an ARM64 environment. Crash when running Detours samples/commem.
![image](https://github.com/microsoft/Detours/assets/6908230/a67a32db-2f80-41d9-b02b-70399ce9e15a)

Is there anything else I need to do to hook COM API of x86,x64 process in ARM64 environment?

**Comment by honkstar1** [comment]
We are having a similar issue atm. Trying to detour an x64 binary when running from arm64 windows. Would love to get advice how to fix or a fix itself :-)

**Comment by kimjw0820** [comment]
> We are having a similar issue atm. Trying to detour an x64 binary when running from arm64 windows. Would love to get advice how to fix or a fix itself :-)

We have been solving the problem by using VMT Hooking method without using Detours in COM API to bypass the problem. However, recently, XFG(Microsoft eXtended Flow Guard) protection API was discovered, and we analyzed that certain APIs cannot be controlled by that method. VMT Hooking method was enough to control general COM API, but we are thinking of a new idea recently.

----------

Microsoft eXtended Flow Guard
eXtended Flow Guard (XFG) has not been officially released yet, but is available in the Windows Insider preview and was publicly presented at Bluehat Shanghai in 2019.[[29]](https://en.wikipedia.org/wiki/Control-flow_integrity#cite_note-Advancing_Windows_Security_Bluehat_Shanghai_2019-29)

XFG extends CFG by validating function call signatures to ensure that indirect function calls are only to the subset of functions with the same signature. Function call signature validation is implemented by adding instructions to store the target function's hash in register r10 immediately prior to the indirect call and storing the calculated function hash in the memory immediately preceding the target address's code. When the indirect call is made, the XFG validation function compares the value in r10 to the target function's stored hash. [[30]](https://en.wikipedia.org/wiki/Control-flow_integrity#cite_note-EXTENDED_FLOW_GUARD_UNDER_THE_MICROSCOPE-30)[[31]](https://en.wikipedia.org/wiki/Control-flow_integrity#cite_note-Exploit_Development:_Between_a_Rock_and_a_(Xtended_Flow)_Guard_Place:_Examining_XFG-31)

https://en.wikipedia.org/wiki/Control-flow_integrity#Microsoft_eXtended_Flow_Guard

**Comment by RatinCN** [comment]
> This may have been fixed with the recent ARM64EC related changes.

Sadly, it seems not work. Here is my test result (using the latest main branch, commit 9764cebcb1a75940e68fa83d6730ffaf0f669401, VS2022):
<img width="975" height="825" alt="Image" src="https://github.com/user-attachments/assets/cafadd12-2c7f-4f63-a802-27f3f2b37a0b" />
Only ARM64 build works. x86, x64, ARM64EC exited with 0xC000001D.

**Comment by mirh** [comment]
That's `STATUS_ILLEGAL_INSTRUCTION`, which makes me wonder about https://github.com/microsoft/Detours/pull/385

**Comment by RatinCN** [comment]
> That's `STATUS_ILLEGAL_INSTRUCTION`, which makes me wonder about [#385](https://github.com/microsoft/Detours/pull/385)

Yes, I've checked on my ARM64 tablet, it writes x64 code to EC code, which is ARM64 instruction, and make it broken.

**Comment by RatinCN** [comment]
I have implemented ARM64EC compatibility in https://github.com/KNSoft/KNSoft.SlimDetours/pull/30. When the hook DLL is built for ARM64EC, it can correctly hook both x64 and ARM64EC target functions in x64 and ARM64EC processes on ARM64 systems. Only some documentation updates remain. I will finish it asap.

**Comment by RatinCN** [comment]
Done, I've fixed this in my [KNSoft.SlimDetours](https://github.com/KNSoft/KNSoft.SlimDetours).
[PR: Support hooking ARM64EC target functions](https://github.com/KNSoft/KNSoft.SlimDetours/pull/30),
[Release v1.2.2-beta](https://github.com/KNSoft/KNSoft.SlimDetours/releases/tag/v1.2.2-beta)
[TechWiki: Implement ARM64EC Compatibility](https://github.com/KNSoft/KNSoft.SlimDetours/blob/main/Docs/TechWiki/Implement%20ARM64EC%20Compatibility/README.md)
I'll also port the fix to Microsoft Detours as soon as possible. However, I’m not sure when the maintainers will respond, as many of my previous PRs have received no response.
